### ISSUE TITLE
Priority rules for GeoKeys conveying similar (and contradictory) information

### ISSUE THREAD

**Issue by rouault** [opengeospatial/geotiff#13]
There are a number of situations where several GeoKeys that can convey implicitly or explicitly an information are found in the same file. A typical example is having ProjectedCSTypeGeoKey which completely defines the CRS, as well as a ProjLinearUnitsGeoKey. In the best case ProjLinearUnitsGeoKey confirms the linear unit of the EPSG code, but sometimes it overrides it with a different unit.

This is not a theoretical concern but issues I've seen in real world : https://trac.osgeo.org/gdal/ticket/6210 and https://trac.osgeo.org/gdal/ticket/4954

Should we allow such override ?

**Comment by EmDevys** [comment]
My opinion is that such override is a real issue in the laxism in TIFF and  even more in GeoTIFF.
IMHO it is total nonsense to override with a different Uom (Unit). what would happen there if one software rely on the standardized unit provided by the CRS definition, and another one on the override !
Just total lack of interoperability, with effects that may be dramatic.
We are supposed to ensure interoperability, and such laxism would drive to lack of interoperability.

**Comment by rouault** [comment]
We could potentially go beyond and completely disallow overrides
Like:
- if ProjectedCSTypeGeoKey != User-Defined, none of Geog*GeoKey and Proj*GeoKey are allowed
- if GeographicTypeGeoKey != User-Defined, none of Geog*GeoKey are allowed
- if GeogEllipsoidGeoKey != User-Defined, GeogSemiMajorAxisGeoKey, GeogSemiMinorAxisGeoKey or GeogInvFlatteningGeoKey are not allowed
- etc..
But that might break files where the information is duplicated but non contradictory.

**Comment by RogerLott** [comment]
This is a general problem with redundent information. Some users want to have redundancy so they can run a check and then invesitigate any discrepancy. GeoTIFF doesn't follow this strategy - it tries to offer the register code as a shortcut.

Although only tangentially relevant to this issue, Topic 2 (18-005r3) clause 7.2 says
"CRS or coordinate operation identification may be through:
a)	a full description, as defined in this document; or 
b)	reference to a full description in a register of geodetic parameters (the reference is made to the register and to the identifier of the object description within that register); or
c)	both a full description and a reference to a full description in a register. If there is a conflict between the two, the object full description should prevail over the reference to a register".

(I would have preferred that the register definition prevailed, but I think it is this way round to cover an inadvertent slip in citing register code value).

Topic 2 goes on to say "a) and b) are alternative means of providing a full description. b) is recommended for simplicity" which is what GeoTIFF does whenever possible. (You would think that if (b) is recommended it would be the one that prevailed if conflict in (c)).

As I understand it, this port is only fixing really broken issues so we simply replicate existing GeoTIFF provisions leave the issue unresolved for now, and fix in a later GeoTIFF revision. Or is this considered to be a breaking issue?

**Comment by cmheazel** [comment]
Our goal is to provide adopters an OGC standard which is equivalent (or as close as possible) to the version that they know and love.  Version 1.0 should only deviate from the 1995 version if there is a compelling need and that need has been documented. Otherwise we may end up with a nice set of shelfware. 

Not saying that there isn't room for improvement. Just that we need to wait until we have sold existing adopters on our version.

**Comment by EmDevys** [comment]
My proposal is rather to keep as is, and leave the issue unresolved, though I would recommend providing a recommendation, based on the best practices (and hopefully rather on use of EPSG register, rather than duplication of information (and the risk of error when doing so). In addition, I would add that we don’t have to be fully consistent with Topic 2 and its c/ usage rule in clause 7.2.

De : RogerLott [mailto:notifications@github.com]
Envoyé : vendredi 21 décembre 2018 18:21
À : opengeospatial/geotiff
Cc : Emmanuel Devys; Comment
Objet : Re: [opengeospatial/geotiff] Priority rules for GeoKeys conveying similar (and contradictory) information (#13)


This is a general problem with redundent information. Some users want to have redundancy so they can run a check and then invesitigate any discrepancy. GeoTIFF doesn't follow this strategy - it tries to offer the register code as a shortcut.

Although only tangentially relevant to this issue, Topic 2 (18-005r3) clause 7.2 says
"CRS or coordinate operation identification may be through:
a) a full description, as defined in this document; or
b) reference to a full description in a register of geodetic parameters (the reference is made to the register and to the identifier of the object description within that register); or
c) both a full description and a reference to a full description in a register. If there is a conflict between the two, the object full description should prevail over the reference to a register".

(I would have preferred that the register definition prevailed, but I think it is this way round to cover an inadvertent slip in citing register code value).

Topic 2 goes on to say "a) and b) are alternative means of providing a full description. b) is recommended for simplicity" which is what GeoTIFF does whenever possible. (You would think that if (b) is recommended it would be the one that prevailed if conflict in (c)).

As I understand it, this port is only fixing really broken issues so we simply replicate existing GeoTIFF provisions leave the issue unresolved for now, and fix in a later GeoTIFF revision. Or is this considered to be a breaking issue?

**Comment by EmDevys** [comment]
I agree for GeoTIFF files adhering to existing specification, dating 1995, with KeyRevision.MinorRevision = 1.0, that should still be handled.

But we should also allow some modernization / correction of the standard, clearly identified in a slightly revised version, which might have KeyRevision.MinorRevision = 1.1, when necessary (and documented).

Does this make sense?


De : Chuck Heazel [mailto:notifications@github.com]
Envoyé : vendredi 21 décembre 2018 18:36
À : opengeospatial/geotiff
Cc : Emmanuel Devys; Comment
Objet : Re: [opengeospatial/geotiff] Priority rules for GeoKeys conveying similar (and contradictory) information (#13)


Our goal is to provide adopters an OGC standard which is equivalent (or as close as possible) to the version that they know and love. Version 1.0 should only deviate from the 1995 version if there is a compelling need and that need has been documented. Otherwise we may end up with a nice set of shelfware.

Not saying that there isn't room for improvement. Just that we need to wait until we have sold existing adopters on our version.

**Comment by thareUSGS** [comment]
*>>But we should also allow some modernization / correction of the
standard, clearly identified in a slightly revised version, which might
have KeyRevision.MinorRevision = 1.1, when necessary (and documented). *

I agree with this and the use of 1.1 as Even has suggested.

-Trent
Trent Hare
Cartographer/GIS Manager
Astrogeology, USGS

**Comment by EmDevys** [comment]
Thanks Martin, this sounds to me like a reasonable proposal, to handle GeoTIFF as is…

De : Martin Desruisseaux [mailto:notifications@github.com]
Envoyé : vendredi 21 décembre 2018 19:25
À : opengeospatial/geotiff
Cc : Emmanuel Devys; Comment
Objet : Re: [opengeospatial/geotiff] Priority rules for GeoKeys conveying similar (and contradictory) information (#13)


The OGC template for standards contains a section about "standardization target types" in its conformance section. Should we said that we have 2 target types?

  *   GeoTIFF writers, which are required to obey to constraints like the ones cited by Even above?
  *   GeoTIFF readers, for which the requirements are relaxed to the level that we have today?

**Comment by EmDevys** [comment]
Should this revision 1.1 add some recommendations on the behaviour of GeoTIFF writers / producers in order to avoid conflicts, and solve conflicts if there are for GeoTIFF readers / users.
Big discussion, seem to be an uneasy issue.
